src/Microsoft.TestPlatform.PlatformAbstractions/common/System/AssemblyResolverUtility.cs

```diff
@@ -0,0 +1,272 @@
+    {
+        private static readonly string[] SupportedFileExtensions = { ".dll", ".exe" };
+
+        private static IPlatformEqtTrace traceImpl = PlatformEqtTrace.Instance;
+
+        /// <summary>
+        /// The directories to look for assemblies to resolve.
+        /// </summary>
+        private HashSet<string> searchDirectories;
+
+        /// <summary>
+        /// Dictionary of Assemblies discovered to date. Must be locked as it may
+        /// be accessed in a multi-threaded context.
+        /// </summary>
+        private Dictionary<string, Assembly> resolvedAssemblies;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AssemblyResolverUtility"/> class.
+        /// </summary>
+        /// <param name="directories"> The search directories. </param>
+        public AssemblyResolverUtility(IEnumerable<string> directories)
+        {
+            this.resolvedAssemblies = new Dictionary<string, Assembly>();
+
+            if (directories == null || !directories.Any())
+            {
+                this.searchDirectories = new HashSet<string>();
+            }
+            else
+            {
+                this.searchDirectories = new HashSet<string>(directories);
+            }
+        }
+
+        /// <summary>
+        /// Set the directories from which assemblies should be searched
+        /// </summary>
+        /// <param name="directories"> The search directories. </param>
+        [System.Security.SecurityCritical]
+        public void AddSearchDirectories(IEnumerable<string> directories)
+        {
+            foreach (var directory in directories)
+            {
+                this.searchDirectories.Add(directory);
+            }
+        }
+
+        /// <summary>
+        /// Assembly Resolve event handler for App Domain - called when CLR loader cannot resolve assembly.
+        /// </summary>
+        /// <param name="assemblyName">
+        /// The assembly Name.
+        /// </param>
+        /// <returns>
+        /// The <see cref="Assembly"/>.
+        /// </returns>
+        public Assembly AssemblyResolverEventHandler(string assemblyName)
+        {
+            if (string.IsNullOrEmpty(assemblyName))
+            {
+                Debug.Assert(false, "AssemblyResolver.OnResolve: args.Name is null or empty.");
+                return null;
+            }
+
+            if (this.searchDirectories == null || this.searchDirectories.Count == 0)
+            {
+                return null;
+            }
+
+            WriteToEqtTrace("AssemblyResolver: {0}: Resolving assembly.", assemblyName);
+
+            // args.Name is like: "Microsoft.VisualStudio.TestTools.Common, Version=[VersionMajor].0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a".
+            lock (this.resolvedAssemblies)
+            {
+                if (this.resolvedAssemblies.TryGetValue(assemblyName, out Assembly assembly))
+                {
+                    if (assembly != null)
```

Comment:
Do we need this check? Will there be any scenario where resolvedAssemblies dictionary will have null value for a assembly name?